### PR TITLE
Add docstrings with Base/Docile syntax.

### PR DESCRIPTION
**Pull request by hayd**
This is initial work towards docstrings, these work on 0.4 (I've tried to ensure every method has a docstring). They should work on 0.3 with docile installed... at least that's my understanding\* - I haven't tested this yet.

On 0.3 without Docile there won't be any docstrings, and the strings will simply be optimized away.

\*If so I can remove the three or four uses of LaTeX, it's only these that require the `doc"..."`.

~~I would like to get Lexicon api generation up and running, but haven't quite worked out the ins and outs yet.~~ The markdown documentation is now generated in doc/build.jl (and, of course, can be tweaked).

### CODE REVIEW

File: src/core.jl [hunk 1]
```diff
@@ -12,9 +12,19 @@ if VERSION < v"0.4.0-dev+4103"
     reverse(p::Pair) = Pair(p.second, p.first)
 end
 
+if VERSION < v"0.4.0-dev"
+    try
+        import Docile: @doc_str
+    catch
+        macro doc_str(x) x end
```
Comment (hayd) [inline src/core.jl]: This is a little hacky, to get the `doc".."` (required only for LaTeX in Markdown).

**Comment by hayd** [comment]
I pushed a commit so that all the markdown files are generated from `doc/build.jl`.

**Comment by sbromberger** [comment]
I can't seem to test it without merging, so let's just go ahead and do it :) @hayd - thank you so much for all the work on this.

**Comment by hayd** [comment]
Lol, great!

Like I say we can tweak the output - the stuff that's generated from methods `{{degree, indegree}}`, I'm sure it could be improved with some thought.

**Comment by hayd** [comment]
IMHO it's looking pretty good

http://lightgraphsjl.readthedocs.org/en/latest/pathing/

**Comment by sbromberger** [comment]
How does this work, exactly? (I noticed a link error in that URL you sent - it's showing `Performs Base.Markdown.Link(Any["Dijkstra's algorithm"],"http://en.wikipedia.org/wiki/Dijkstra%27s_algorithm")`) and I thought it might be a good way to learn how to make adjustments.

**Comment by hayd** [comment]
One annoying thing is that the LHS doesn't seem to show a difference between h2 and h3:

nested

**Comment by hayd** [comment]
Good spot, I think that's a bug (which was fixed recently in master), missing plain method for Link.

If you're running a very recent julia, just run `julia doc/build.jl` and if the diff is fixed.

**Comment by sbromberger** [comment]
OK, did that, and committed directly to master. Let's see what happens :)

**Comment by sbromberger** [comment]
Hm. Links work, but the prototypes are now missing.

**Comment by hayd** [comment]
This is weird. I wonder if you need Lexicon installed on your system? cc @MichaelHatherly

(Always check the diff!!!!)

**Comment by sbromberger** [comment]
easy enough to revert - but let's see if we can get both things nailed :)

**Comment by hayd** [comment]
Can you try `Pkg.add("Lexicon")` then redo it?

**Comment by sbromberger** [comment]
Didn't result in any changes.

**Comment by hayd** [comment]
@sbromberger do any help messages work for you? i.e. if you do `using LightGraphs` `?LightGraphs` in the repl. Updating my base... very strange.

**Comment by sbromberger** [comment]
The help messages work in the REPL....

```
help?> is_directed
search: is_directed

  Returns true if g is a DiGraph.
```

**Comment by hayd** [comment]
Hmmmm, now I think about it, it's not the help that's an issue. It's the method table. Are you working on LightGraphs on a different place to it's installed (i.e. not in .julia/v0.4/LightTables) ?

Part of this code checks that the build file is in the same location-ish as the method definitions.

File: doc/build.jl [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+    println(io, "```")
+    for m in methods(f)
+        md_method(io, m)
+    end
+    println(io, "```")
+end
+function md_method(io, m)
+    # We only print methods with are defined in the parent (project) directory
+    if !(startswith(string(m.func.code.file),
+                    splitdir(dirname(@__FILE__))[1]))
```
Comment (hayd) [inline doc/build.jl]: This bit. If your LightTables is being imported from a different place that's gonna fail.

We could test the file of LightTables, in fact that's probably better than this!

**Comment by sbromberger** [comment]
Oh - yes. (assuming you meant LightGraphs instead of LightTables) - ~/.julia/v0.4/LightGraphs is symlinked to a code directory I have.

**Comment by hayd** [comment]
lol, yes. Do you know how to get the actual file/directory of a Module (so we can fix this that way)?

**Comment by sbromberger** [comment]
hm. Does something like `@__FILE__` do anything for you?

**Comment by hayd** [comment]
Any idea why the realpath doesn't work? Perhaps add some print statements (for the `@___FILE__` and the `string(m.func.code.file)` and see if anything jumps out?

**Comment by sbromberger** [comment]
I added two lines to build.jl:

```
function md_method(io, m)
info("file = $(@__FILE__)")
info("code.file = $(m.func.code.file)")
...

```

Got this partial result:
```
 - 'basicmeasures.md'
INFO: file = /Users/seth/dev/julia/wip/LightGraphs.jl/doc/build.jl
INFO: code.file = /Users/seth/.julia/v0.4/LightGraphs/src/core.jl
INFO: file = /Users/seth/dev/julia/wip/LightGraphs.jl/doc/build.jl
INFO: code.file = /Users/seth/.julia/v0.4/LightGraphs/src/core.jl
```

**Comment by hayd** [comment]
what's is `realpath("/Users/seth/dev/julia/wip/LightGraphs.jl/doc/build.jl")` ?

**Comment by sbromberger** [comment]
From the REPL?

```
julia> realpath("/Users/seth/dev/julia/wip/LightGraphs.jl/doc/build.jl")
"/Users/seth/dev/julia/wip/LightGraphs.jl/doc/build.jl"
```

**Comment by hayd** [comment]
I'd hoped that would resolve symlinks? I'm very confused.

**Comment by sbromberger** [comment]
That's not a symlink - that's the real path. I have

`/Users/seth/.julia/v0.4/LightGraphs -> /Users/seth/dev/julia/wip/LightGraphs.jl`

**Comment by hayd** [comment]
So we just realpath them both?

**Comment by hayd** [comment]
```
+    if !(startswith(realpath(string(m.func.code.file)),
+                    splitdir(dirname(realpath(@__FILE__)))[1]))
```

**Comment by sbromberger** [comment]
```
julia> realpath("/Users/seth/.julia/v0.4/LightGraphs")
"/Users/seth/dev/julia/wip/LightGraphs.jl"
```

**Comment by sbromberger** [comment]
why not just `dirname(@__FILE__)` if you're looking for the script's directory? That's how we're doing it in `test/persistence.jl`.

**Comment by hayd** [comment]
That's `LightGraphs.jl/doc`, you need the parent `LightGraphs.jl/`. And this could be simlinked the other way... :p so I think the realpath is harmless/desired.

?

**Comment by sbromberger** [comment]
(see edit to previous comment)

**Comment by hayd** [comment]
That's cleaner. As long as your realpathing both of them, we're in the clear. Is the build script working for you now?

**Comment by sbromberger** [comment]
Ah, I haven't actually made the change yet :) let me try.

**Comment by sbromberger** [comment]
Should I hard-code Pkg.dir("LightGraphs") then, or is there another way to get the package name?

**Comment by hayd** [comment]
I couldn't find another way aside from the hack above. I'm easy.

Hmm, there is (small) buggyness here though https://lightgraphsjl.readthedocs.org/en/latest/operators/ :(

**Comment by sbromberger** [comment]
OK, I have no idea where to put this code.

**Comment by sbromberger** [comment]
See https://github.com/JuliaGraphs/LightGraphs.jl/pull/112

**Comment by sbromberger** [comment]
Another issue is that typealiases appear to be expanded in the docs: For example, the code has

```
typealias Edge Pair
...

dst(e::Edge) = e.second
```

but the method documentation produces

```
dst(e::Pair{Int64,Int64})
```

It would be much better if this followed the Edge abstraction instead of providing the internal structure, especially since it's different for 0.3 (0.3 doesn't have Pair).

**Comment by hayd** [comment]
Hmmm... not sure how to pull that out.
```
julia> methods(dst)
# 1 method for generic function "dst":
dst(e::Pair{Int64,Int64}) at /Users/andy/.julia/v0.4/LightGraphs/src/core.jl:28
```
:s

**Comment by sbromberger** [comment]
I'm wondering whether not including the types is the better approach. That is, instead of

```
dst(e::Pair{Int64,Int64})
```
in the docs, we just have

```
dst(e)
```
since the text itself explains the significance of the parameters.

**Comment by hayd** [comment]
I think what you want to do is replace type aliases as the last think when printing... will think on it.

**Pull request merged**